deltacat/compute/compactor_v2/steps/merge.py

```diff
@@ -1,18 +1,439 @@
+            input.hash_group_index, input.hash_bucket_count, input.num_hash_groups
+        )
+
+        total_deduped_records = 0
+        total_dfes_found = 0
+
+        materialized_results: List[MaterializeResult] = []
+        hb_index_copy_by_reference = []
+        for hb_idx in valid_hb_indices_iterable:
+            dfe_list = hb_index_to_delta_file_envelopes_list.get(hb_idx)
+
+            if dfe_list:
+                total_dfes_found += 1
+                table = _build_incremental_table(hb_idx, dfe_list)
+
+                incremental_len = len(table)
+                logger.info(
+                    f"Got the incremental table of length {incremental_len} for hash bucket {hb_idx}"
+                )
+
+                if input.sort_keys:
+                    # Incremental is sorted and merged, as sorting
+                    # on non event based sort key does not produce consistent
+                    # compaction results. E.g., compaction(delta1, delta2, delta3)
+                    # will not be equal to compaction(compaction(delta1, delta2), delta3).
+                    table = table.sort_by(input.sort_keys)
+
+                compacted_table = None
+                if (
+                    input.round_completion_info
+                    and input.round_completion_info.hb_index_to_entry_range
+                    and input.round_completion_info.hb_index_to_entry_range.get(
+                        str(hb_idx)
+                    )
+                    is not None
+                ):
+
+                    compacted_table = _download_compacted_table(
+                        hb_index=hb_idx,
+                        rcf=input.round_completion_info,
+                        read_kwargs_provider=input.read_kwargs_provider,
+                        deltacat_storage=input.deltacat_storage,
+                        deltacat_storage_kwargs=input.deltacat_storage_kwargs,
+                    )
+
+                hb_table_record_count = len(table) + (
+                    len(compacted_table) if compacted_table else 0
+                )
+
+                table, merge_time = timed_invocation(
+                    func=_merge_tables,
+                    table=table,
+                    primary_keys=input.primary_keys,
+                    can_drop_duplicates=input.drop_duplicates,
+                    compacted_table=compacted_table,
+                )
+                total_deduped_records += hb_table_record_count - len(table)
+
+                logger.info(
+                    f"[Merge task index {input.merge_task_index}] Merged "
+                    f"record count: {len(table)}, took: {merge_time}s"
+                )
+
+                materialized_results.append(_materialize(hb_idx, [table]))
+            else:
+                hb_index_copy_by_reference.append(hb_idx)
+
+        if input.round_completion_info and hb_index_copy_by_reference:
+            referenced_materialized_results = (
```

Comment:
Can we add logging during reference copy as well?

Reply:
Sure.